sm/sm.c

```diff
@@ -413,7 +413,8 @@ static void state_set(struct m0_sm *mach, int state, int32_t rc)
 {
 	const struct m0_sm_state_descr *sd;
 
-	mach->sm_rc = rc;
+	if (!(strcmp(mach->sm_conf->scf_name, "btree-conf") == 0))
+		mach->sm_rc = rc;
```

Comment:
No this is not correct. The state machine implementation should not be partial to any module (such as btree here).

Reply:
Ok.
Then I am not sure how do we get the return code.
Because till this point, mach->sm_rc has our error code. After the assignment on line 417, our error code vanishes as rc = 0.

Reply:
As discussed use the patch and the change which I have requested in the PR for adding M0_EXPORTED attribute to the new sm function and that should be enough.